src/main/flight/gps_rescue.c

```diff
@@ -669,25 +672,28 @@ void descend(void)
         // if quad drifts further than 2m away from home, should by then have rotated towards home, so pitch is allowed
-        rescueState.intent.rollAngleLimitDeg = 0.5f * gpsRescueConfig()->maxRescueAngle * rescueState.intent.proximityToLandingArea;
-        // reduce roll capability when closer to home, none within final 2m
-        rescueState.intent.pitchAngleLimitDeg = gpsRescueConfig()->maxRescueAngle * rescueState.sensor.groundspeedPitchAttenuator;
+        rescueState.intent.rollAngleLimitDeg = 0.5f * rescueState.intent.pitchAngleLimitDeg * (rescueState.sensor.distanceToHomeM / rescueState.intent.descentDistanceM);
+        // reduce roll capability when closer to home, keeping a little bit within final 2m to prevent endless yaw spin at 2m
+        // limit roll angle from GPS Rescue code to half the max pitch angle; earth referenced yaw may add more roll via angle code
     }
 
     rescueState.intent.yawAttenuator = 1.0f; // just in case entered descend phase before yaw authority was complete
 
-    // configure altitude step for descent, considering interval between altitude readings
-    rescueState.intent.altitudeStep = -1.0f * rescueState.sensor.altitudeDataIntervalSeconds * gpsRescueConfig()->descendRate;
+    rescueState.intent.altitudeStep = rescueState.sensor.altitudeDataIntervalSeconds * gpsRescueConfig()->descendRate;
+    // set the altitude step, considering the interval between altitude readings and the descent rate
 
-    // descend more slowly if return altitude is less than 20m
-    const float descentAttenuator = rescueState.intent.returnAltitudeCm / 2000.0f;
-    if (descentAttenuator < 1.0f) {
-        rescueState.intent.altitudeStep *= descentAttenuator;
-    }
-    // descend more quickly from higher altitude
-    rescueState.intent.descentRateModifier = constrainf(rescueState.intent.targetAltitudeCm / 5000.0f, 0.0f, 1.0f);
-    rescueState.intent.targetAltitudeCm += rescueState.intent.altitudeStep * (1.0f + (2.0f * rescueState.intent.descentRateModifier));
-    // increase descent rate to max of 3x default above 50m, 2x above 25m, 1.2 at 5m, default by ground level.
+    const float descentRateAttenuator = constrainf (rescueState.intent.returnAltitudeCm / 2000.0f, 0.25f, 1.0f);
+    rescueState.intent.altitudeStep *= descentRateAttenuator;
+    // descend more slowly if the initial return altitude was set to less than 20m
+    // slowest descent rate will be 1/4 of normal when we start descending at or below 5m above take-off point.
+    // otherwise a rescue initiated very low and close may not get all the way home
+
+    rescueState.intent.descentRateBooster = constrainf(rescueState.intent.targetAltitudeCm / 5000.0f, 0.0f, 1.0f);
+    // increase descent rate to max of 3x default above 50m, 2x above 25m, 1.2 at 5m, default by ground level
+    // also boosts altitude D for better control at higher descent rates
+    rescueState.intent.altitudeStep *= 1.0f + (2.0f * rescueState.intent.descentRateBooster); 
```

Comment:
```suggestion
    const float descentRateBooster = constrainf(rescueState.intent.targetAltitudeCm / 5000.0f, 0.0f, 1.0f);
    // increase descent rate to max of 3x default above 50m, 2x above 25m, 1.2 at 5m, default by ground level
    // also boosts altitude D for better control at higher descent rates
    rescueState.intent.altitudeStep *= 1.0f + (2.0f * descentRateBooster); 
    rescueState.intent.throttleDGain = 1.0f + rescueState.intent.throttleDGain;
```

Reply:
I don't follow the `rescueState.intent.throttleDGain` refactoring concept. The `descentRateBooster` is something that increases the descent speed according to the height above target.  This works well because we can have a relatively low terminal descent speed when close to ground, but descend faster while high up.  Later I found that at the start of the descent, where we get a sudden change from zero descent speed to a rapid descent speed, it was helpful to increase the D element (since we get D from error) to minimise the lag in the early phase of the descent.  So I used the same factor to effectively double the D factor at the start, especially if high and descending fast.  The amount of D boost required at lower altitude was less because the descent rate, and the change in descent rate, was less.  
So this `descentRateBooster` may be badly named, but it is quite different from the throttleDGain.

Reply:
descentRateBooster is used in altitude pid as `throttleD = gpsRescueConfig()->throttleD * verticalSpeed * (1 + descentRateBooster);`

The idea is to rename it to dGain when related to PID and calculate it here. In PID code, it it not that important to know from where is D gain coming from. It is simply PID.
Here, boosting PID D gain is relevant. So set `throttleDGain` (it is applied to user-specified `gpsRescueConfig()->throttleD`, keep naming consistent). But don't worry about PID internals.

Another use is altitudeStep adjustment. Just do it here, but don't mix these two parameters in other places.

Reply:
Done, I think, in the way you suggested